bundles/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/internal/KM200Device.java

```diff
@@ -0,0 +1,257 @@
+            } catch (NoSuchAlgorithmException e) {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
+            }
+
+            /* First half of the key: MD5 of (GatewayPassword . Salt) */
+            try {
+                bytesOfGatewayPassword = gatewayPassword.getBytes("UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
+            }
+            byte[] CombParts1 = new byte[bytesOfGatewayPassword.length + MD5Salt.length];
+            System.arraycopy(bytesOfGatewayPassword, 0, CombParts1, 0, bytesOfGatewayPassword.length);
+            System.arraycopy(MD5Salt, 0, CombParts1, bytesOfGatewayPassword.length, MD5Salt.length);
+            MD5_K1 = md.digest(CombParts1);
+
+            /* Second half of the key: - Initial: MD5 of ( Salt) */
+            MD5_K2_Init = md.digest(MD5Salt);
+
+            /* Second half of the key: - private: MD5 of ( Salt . PrivatePassword) */
+            try {
+                bytesOfPrivatePassword = privatePassword.getBytes("UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
+            }
+            byte[] CombParts2 = new byte[bytesOfPrivatePassword.length + MD5Salt.length];
+            System.arraycopy(MD5Salt, 0, CombParts2, 0, MD5Salt.length);
+            System.arraycopy(bytesOfPrivatePassword, 0, CombParts2, MD5Salt.length, bytesOfPrivatePassword.length);
+            MD5_K2_Private = md.digest(CombParts2);
+
+            /* Create Keys */
+            cryptKeyInit = new byte[MD5_K1.length + MD5_K2_Init.length];
+            System.arraycopy(MD5_K1, 0, cryptKeyInit, 0, MD5_K1.length);
+            System.arraycopy(MD5_K2_Init, 0, cryptKeyInit, MD5_K1.length, MD5_K2_Init.length);
+
+            cryptKeyPriv = new byte[MD5_K1.length + MD5_K2_Private.length];
+            System.arraycopy(MD5_K1, 0, cryptKeyPriv, 0, MD5_K1.length);
+            System.arraycopy(MD5_K2_Private, 0, cryptKeyPriv, MD5_K1.length, MD5_K2_Private.length);
+
+        }
+    }
+
+    // getter
+    public String getIP4Address() {
+        return ip4Address;
+    }
+
+    public String getGatewayPassword() {
+        return gatewayPassword;
+    }
+
+    public String getPrivatePassword() {
+        return privatePassword;
+    }
+
+    public byte[] getCryptKeyInit() {
+        return cryptKeyInit;
+    }
+
+    public byte[] getCryptKeyPriv() {
+        return cryptKeyPriv;
+    }
+
+    public String getCharSet() {
+        return charSet;
+    }
+
+    public Boolean getInited() {
+        return inited;
+    }
+
+    /**
+     * This function outputs a ";" separated list of all on the device available services with its capabilities
+     *
+     * @author Markus Eckhardt
+     *
+     * @since 1.9.0
+     */
+    public void listAllServices() {
```

Comment:
How often is this method called? If it is (or can be) called frequently, these logger statements should probably be debug instead of info.